### DISCUSSION TITLE
why is there a migrations container in the sample docker compose

### DISCUSSION THREAD

**Discussion by jmls** [ory/kratos#3115] [Ideas]
I'd like to understand why there is a container for migrations in the sample docker compose, and why kratos depends on it. 

The docs state this

![image](https://user-images.githubusercontent.com/929523/217602091-1e3673cf-0999-4f3c-9a9b-c7d44f50c561.png)

but - if  "**_Migrations also run automatically whenever you run the serve command_**" then why is the migrations container needed ?

**Comment by vinckr** [comment]
Hello @jmls 

The kratos-migrate container in the Docker Compose file is designed to handle SQL migrations, which are necessary for new installations and when upgrading. This container is set up to run the migrations before the main kratos service starts, ensuring that the database schema is up-to-date before the application begins to interact with it.
While it's true that migrations also run automatically whenever you run the serve command, having a separate container for migrations can be beneficial for several reasons:
1. Separation of concerns: It keeps the migration process separate from the main application process. This can make it easier to manage and troubleshoot issues related to database migrations.
1. Startup dependencies: Docker Compose allows you to specify that certain services depend on others, meaning they won't start until the services they depend on have started. In this case, the kratos service won't start until the kratos-migrate service has completed, ensuring that the database is ready before the application starts.
1. Parallelization: If you have multiple services that depend on the database, they can all wait for the kratos-migrate service to complete, then start in parallel. This can potentially speed up the startup process.
